Sources/swift-format/Utilities/Helpers.swift

```diff
@@ -42,16 +42,33 @@ func readSource(from fileHandle: FileHandle) -> String? {
 func processSources(
   from paths: [String], configurationPath: String?,
   diagnosticEngine: DiagnosticEngine,
-  transform: (FileHandle, String, Configuration) -> Void
+  transform: @escaping (FileHandle, String, Configuration) -> Void
 ) throws {
+  let concurrentQueue = DispatchQueue(label: "com.queue.Concurrent", qos: .userInitiated, attributes: .concurrent)
+  let group = DispatchGroup()
+  let lock = NSLock()
+  var lastError: Error?
   for path in FileIterator(paths: paths) {
-    guard let sourceFile = FileHandle(forReadingAtPath: path) else {
-      diagnosticEngine.diagnose(
-        Diagnostic.Message(.error, "Unable to create a file handle for source from \(path)."))
-      return
+    concurrentQueue.async(group: group) {
+      guard let sourceFile = FileHandle(forReadingAtPath: path) else {
+        diagnosticEngine.diagnose(
+          Diagnostic.Message(.error, "Unable to create a file handle for source from \(path)."))
+        return
+      }
+      do {
+        let configuration = try loadConfiguration(forSwiftFile: path, configFilePath: configurationPath)
+        transform(sourceFile, path, configuration)
+      } catch let error {
+        lock.lock()
+        lastError = error
+        lock.unlock()
+      }
     }
-    let configuration = try loadConfiguration(forSwiftFile: path, configFilePath: configurationPath)
-    transform(sourceFile, path, configuration)
+  }
+
+  group.wait()
+  if let error = lastError {
+    throw error
```

Comment:
Up for suggestions on different error handling options here

Reply:
Thanks for continuing with this in the face of constant refactoring 😅 (Another one of which is coming down the pipe: https://github.com/apple/swift-format/pull/178)

Taking a closer look at the original code (before your PR), its intended behavior was unclear to me (because it's an edge case that we've never hit). If we failed to open a source file, we'd emit a diagnostic and then return. If an error is thrown trying to read the config file for a particular file, we let it get thrown passed us, letting the argument parser's logic catch it. In either case, an error at any time stops linting/formatting for the whole file set.

Spot-checking what some other formatters do in similar situations:

* Prettier [logs an error message and continues on](https://github.com/prettier/prettier/blob/master/src/cli/util.js#L456-L464) if it can't open the file.
* Clang-format [logs an error message and continues if it fails to open a source file](https://github.com/llvm/llvm-project/blob/master/clang/tools/clang-format/ClangFormat.cpp#L350-L353) or if there's a [problem getting the style](https://github.com/llvm/llvm-project/blob/master/clang/tools/clang-format/ClangFormat.cpp#L382-L385), among other conditions.

So our current code was making this a lot harder than it needed to be, and https://github.com/apple/swift-format/pull/178 cleans all that up (among other things) so now instead of worrying about early exit or throwing conditions, a bad file/configuration just causes a diagnostic to be emitted and we move on, and you don't have to think about the error conditions at all.

---

Aside from that, I was thinking that it would be nice to control this with a `-j <N>/--jobs=<N>` flag so that we can try rolling it out off by default at first and then flip it on later once we've put it through some stress tests. But AFAIK Dispatch makes it kind of hard to control that. Would it make sense to use `OperationQueue` here instead?

Reply:
Personally, I much prefer a `--parallel` (or equivalent name) flag instead of `--jobs=N` because I'm never sure how many to pick. I think it would be great if it just let `Dispatch` decide how many jobs instead of letting the user control it :)

Reply:
That would work too. I don't have too strong an opinion here about the specifics, as long as it's off by default (for now) and doesn't cause excessive thread creation when it's enabled.

Maybe each format/lint job is short-running enough that we don't need to worry about finer grained control, or we can start with a binary `--parallel` flag now and add job count control later if we find we need it.

Reply:
Added the `parallel` flag!